[InstrFDO]Allow indexed profile reader to parse compatible future versions and returns errors for incompatible ones.

mingmingl-llvm opened the pull request:
_(We use profile reader to refer to indexed profile reader for simplicity.)_ 

Profile reader won't parse profiles of a future version ([code](https://github.com/llvm/llvm-project/blob/2f2e31c3c980407b2660b4f5d10e7cdb3fa79138/llvm/lib/ProfileData/InstrProf.cpp#L1616-L1618)) currently. This can cause usability pains (elaborated below)

Profiles of a new version come in two forms in terms of its compatibility with profile readers compiled at old commit (that is not taught to parse this new version):
* Case 1: Semantics of existing payload sections remain unchanged. New header fields and new payload sections are added. 
* Case 2: Semantics of existing sections might change in a backward compatible way. For instance, reader dispatches to different code path depending on the profile version or sub-section version.

The motivation of this PR is to allow profile readers to _make use of_ profiles for case 1 and retain the capability to fail early for case 2. For case 1, both `llvm-profdata` (for profile processing) and compilers (for optimized build) are covered.
* To assist the profile reader to detect when the reader is not compatible with (future versions of) profile, the profile records `MinCompatibleReaderVersion`.
* With this change, profile reader can locate the end byte offset of header from profiles (i.e., not depending on the header C++ struct definition).

_Usability pains_
 - Currently, the indexed profile reader relies on the C++ struct [definition](https://github.com/llvm/llvm-project/blob/d97d560fbf6ed26a198b3afe1594d7d63b88ab3a/llvm/include/llvm/ProfileData/InstrProf.h#L1158-L1191) of Header [to know the byte size of the profile header](https://github.com/llvm/llvm-project/blob/d97d560fbf6ed26a198b3afe1594d7d63b88ab3a/llvm/lib/ProfileData/InstrProf.cpp#L1648-L1672), which is baked in the binary.  Thereby older compilers won't be able to parse profiles correctly if any new field is added (for new payload sections) in the header C++ struct.
- In the real world, users might profile their instrumented binaries on a server farm and use orchestration services to merge profiles, and then feedback the merged profiles for optimized builds to compilers. The release cycle for compilers and orchestration tools might be different and not easy to coordinate.
   - Say a commit `X` introduces a new section `S` and bumps an indexed profile version from `V` to `V+1`.  Without header forward compatibility, things could break in various ways.  For instance, if the orchestration tool generates profiles of `V+1` but compiler release falls behind, the compiler cannot parse the post-processed indexed profiles to generated FDO-optimized binaries. A better choice is to make the compiler parse the known profile sections and ignore the new unknown one to proceed with optimized build.

mingmingl-llvm committed:
[InstrFDO]Record the on-disk header size in indexed profile header

mingmingl-llvm committed:
polish

minglotus-6 commented:
cc @amharc as I somehow cannot add you as a reviewer

minglotus-6 commented on llvm/lib/ProfileData/InstrProf.cpp:
```diff
@@ -1650,10 +1659,12 @@ size_t Header::size() const {
                   "Please update the size computation below if a new field has "
                   "been added to the header, if not add a case statement to "
                   "fall through to the latest version.");
+  case 13ull:
```
Hmm, I just realized this works for version 13 but doesn't initialize 'Header::size'. Working on a fix now.

minglotus-6 replied:
done in this [commit](https://github.com/llvm/llvm-project/pull/88212/commits/3ac8075b1f893e91f4a3bd5495f4494c87b752eb).

mingmingl-llvm committed:
Fix the write and read of 'Size' field

minglotus-6 commented on llvm/lib/ProfileData/InstrProf.cpp:
```diff
@@ -1598,48 +1598,62 @@ uint64_t Header::formatVersion() const {
+  case 13ull:
+    // Size field is already read.
+    FieldByteOffset -= sizeof(Header::Size);
+    [[fallthrough]];
```
Use 'sizeof(Header::<Field>` here and below for better readability.

Yet still, this implicitly depends on the fact header fields are written out as type `uint64_t` (both in-memory and on-disk) as [`OS::write`](https://github.com/llvm/llvm-project/blob/36e25772ddd049c8c742e55fbd2b3c9aaceb7060/llvm/lib/ProfileData/InstrProfWriter.cpp#L609-L637) used by indexed profile writer only writes `uint64_t` for integers.

mingmingl-llvm committed:
Update FieldByteOffset to be the last known header field that's understandable by the binary

mingmingl-llvm committed:
Allow unknown future versions in profile reader.

- Before this, profile reader check and rejects unknown future versions.
  This prevents unknown errors when the semantics change in an existing
  payload section.
- After this change, we want to allow profile reader to skip unknown new
  sections but still prevent unknown errors upon semantic changes.
- One middle ground is to record the compatible versions in the
  profiles. Profile readers can read back the compatible versions and
  rejects a profile if the intersection between reader-supported version
  and profile-compatible version is empty. Still working on finalizing
  the solution.

mingmingl-llvm committed:
Add Header::MinimumProfileReaderVersion to detect incompatible changes

mingmingl-llvm committed:
run 'git merge main' and resolve conflict

minglotus-6 commented on llvm/lib/ProfileData/InstrProfWriter.cpp:
```diff
@@ -710,9 +745,28 @@ Error InstrProfWriter::writeImpl(ProfOStream &OS) {
+  // This is a test-only path to append dummy header fields.
+  // NOTE: please write all other header fields before this one.
+  if (AppendAdditionalHeaderFields)
     OS.write(0);
```
I wonder if I should move this to a member function `InstrProfWriter::TestOnlyAppendAdditionalHeaaderFields` so people don't need to carefully write real header fields before this line (or spend time debugging what goes wrong).

An alternative (that i'm leaning towards) is to introduce `InstrProfWriter::writeHeader` method to encapsulate the write of header, but with back patching (`InstrProfWriter::writeHeader` records the byte offset to patch after payloads are written), it's a bit of refactoring work so decided to ask for thoughts first.

snehasish replied:
cc: @kazutakahirata who also suggested a refactoring of the instrprof writer code. In general I think that's a good direction to breakdown the monolithic implementation which exists today.

minglotus-6 replied:
sounds good. I sent out https://github.com/llvm/llvm-project/pull/90142 to add `InstrProfWriter::writeHeader` and will make it a parent of this patch.

mingmingl-llvm committed:
Record the profile byte size and temporal profile byte size, such that reader can locate the end of the profile and the end o of the temporal profile payload

mingmingl-llvm committed:
Read new header fields in the correct endian; and update the calculatioon of temporal profile end byte

minglotus-6 commented:
The precommit CI test failure should be irrelevant

```
# github-pull-requests_build_59610_linux-linux-x64.log is a download copy from https://buildkite.com/llvm-project/github-pull-requests/builds/59610#018f2717-df24-41e0-afaa-c455f02b3127
 
cat github-pull-requests_build_59610_linux-linux-x64.log | grep -B 5 -A 10 "Failed Tests"
[ 4.0s, 6.0s) :: [                                        ] :: [   207/103045]
[ 2.0s, 4.0s) :: [                                        ] :: [   967/103045]
[ 0.0s, 2.0s) :: [*************************************** ] :: [101704/103045]
--------------------------------------------------------------------------
********************
Failed Tests (1):
  BOLT :: RISCV/unnamed-sym-no-entry.c


Testing Time: 281.03s

Total Discovered Tests: 116456
  Skipped          :     73 (0.06%)
  Unsupported      :   2801 (2.41%)
  Passed           : 113283 (97.28%)
  Expectedly Failed:    298 (0.26%)
--
[0.20s,0.30s) :: [*******                                 ] :: [ 81/419]
[0.10s,0.20s) :: [****************                        ] :: [172/419]
[0.00s,0.10s) :: [*******                                 ] :: [ 82/419]
--------------------------------------------------------------------------
********************
Failed Tests (1):
  BOLT :: RISCV/unnamed-sym-no-entry.c


Testing Time: 1.93s

Total Discovered Tests: 431
  Skipped          :   7 (1.62%)
  Unsupported      :  13 (3.02%)
  Passed           : 409 (94.90%)
  Expectedly Failed:   1 (0.23%)
```

mingmingl-llvm committed:
Changes
1. merge main branch
2. reduce the number of new header fields from 4 to 2.
3. polish comments

minglotus-6 commented:
With the clean-ups like https://github.com/llvm/llvm-project/pull/93346 and https://github.com/llvm/llvm-project/pull/93594, this patch is much simpler than its previous version (thanks to `readNext` and back patching once). PTAL, thanks!

kazutakahirata reviewed (commented):
I think the on-disk header size is very useful.  People can keep adding new sections without affecting others who are not interested in the new sections.

I am not to sure about the minimum compatible version.  I have a couple of concerns:

- When somebody bumps the minimum compatible version due to an update to one specific section, they would teach `InstrProfWriter::minCompatibleReaderVersion` to return `Version14` or some higher number.  This is a profile-wide number, not the version of some specific section.  This means that users lose the forward compatibility regardless of whether they are using the updated section.
- The complexity is another concern.  You and I have discussed the min compatible version in person, so I understand it.  However, when somebody comes along and wishes to add or modify a section, it's a challenge to understand how it works.  Cursory reading of the source code most likely would not help them answer questions like "I'm modifying an existing section.  Should I add that as a new section?  Or should I modify the existing section and update the min compatible version?"

In any event, there should be some documentation, either as comments in the code or some `.rst`, about how to add and modify a section.  Relevant functions and variables should point to the documentation.

minglotus-6 commented:
> I think the on-disk header size is very useful. People can keep adding new sections without affecting others who are not interested in the new sections.
> 
> I am not to sure about the minimum compatible version. I have a couple of concerns:
> 
> * When somebody bumps the minimum compatible version due to an update to one specific section, they would teach `InstrProfWriter::minCompatibleReaderVersion` to return `Version14` or some higher number.  This is a profile-wide number, not the version of some specific section.  This means that users lose the forward compatibility regardless of whether they are using the updated section.

Currently many payload sections don't have a sub-section version and indexed profile format doesn't have any forward compatibility, I think introducing profile-wide min compatible version is simpler than backfilling versions for every payload sections, what do you think?

I agree it's a valid use case to update version of specific payload sections. If the payload section have versions (e.g., memory profiles), profile reader can do the version check on sub-sections and there is no need to bump profile-wide versions.

>  In any event, there should be some documentation, either as comments in the code or some .rst, about how to add and modify a section. Relevant functions and variables should point to the documentation.

Good point about documentation update. Comments have closer proximity to code for details but doc can summarize high-level ideas. I have https://llvm.org/docs/InstrProfileFormat.html an https://clang.llvm.org/docs/SourceBasedCodeCoverage.html#format-compatibility-guarantees in mind for now. Will update the pr.

mingmingl-llvm committed:
commit code review suggestions in batch

Co-authored-by: Kazu Hirata <kazu@google.com>

mingmingl-llvm committed:
apply suggestions in batch.

Co-authored-by: Kazu Hirata <kazu@google.com>
